build_runner/lib/src/entrypoint/run_script.dart

```diff
@@ -0,0 +1,145 @@
+      '<executable> [-- [script-arguments]]';
+
+  @override
+  SharedOptions readOptions() {
+    // The default option parser will throw if we pass additional arguments,
+    // because it expects positional arguments to be build directories.
+    //
+    // Instead, pass an empty list, and we'll handle positional arguments ourselves.
+    return new SharedOptions.fromParsedArgs(
+        argResults, [], packageGraph.root.name, this);
+  }
+
+  @override
+  FutureOr<int> run() async {
+    // Ensure that the user passed the name of a file to run.
+    if (argResults.rest.isEmpty) {
+      stderr
+        ..writeln('Must specify an executable to run.')
+        ..writeln()
+        ..writeln(usage);
+      return ExitCode.usage.code;
+    }
+
+    var scriptName = argResults.rest[0];
+    var passedArgs = argResults.rest.skip(1).toList();
+    var options = readOptions();
+
+    // Create a temporary directory in which to execute the script.
+    var tempPath = Directory.systemTemp
+        .createTempSync('build_runner_run_script')
+        .absolute
+        .uri
+        .toFilePath();
+
+    // Create two ReceivePorts, so that we can quit when the isolate is done.
+    //
+    // Define these before starting the isolate, so that we can close
+    // them if there is a spawn exception.
+    ReceivePort onExit, onError;
+
+    try {
+      var outputMap = options.outputMap ?? {};
+      outputMap.addAll({tempPath: null});
+
+      var result = await build(
+        builderApplications,
+        deleteFilesByDefault: options.deleteFilesByDefault,
+        enableLowResourcesMode: options.enableLowResourcesMode,
+        configKey: options.configKey,
+        assumeTty: options.assumeTty,
+        outputMap: outputMap,
+        packageGraph: packageGraph,
+        verbose: options.verbose,
+        builderConfigOverrides: options.builderConfigOverrides,
+        isReleaseBuild: options.isReleaseBuild,
+        trackPerformance: options.trackPerformance,
+        skipBuildScriptCheck: options.skipBuildScriptCheck,
+        buildDirs: options.buildDirs,
+        logPerformanceDir: options.logPerformanceDir,
+      );
+
+      if (result.status == BuildStatus.failure) {
+        stdout.writeln('Skipping script run due to build failure');
+        return result.failureType.exitCode;
+      }
+
+      // Find the path of the script to run.
+      var scriptPath = p.setExtension(p.join(tempPath, scriptName), '.dart');
+      var packageConfigPath = p.join(tempPath, '.packages');
+
+      // Use a completer to determine the exit code.
+      var completer = new Completer<int>();
+
+      Isolate isolate;
+      onExit = new ReceivePort();
+      onError = new ReceivePort();
+
+      // On an error, kill the isolate, and rethrow the error.
+      onError.listen((e) {
+        isolate?.kill();
+        onExit.close();
+        onError.close();
+        Zone.current.handleUncaughtError(
```

Comment:
I think we probably just want to `_logger.severe('Unhandled error from script: $scriptName', e[0], e[1])` instead of rethrowing it (which is effectively what this does I think?).